File: keyczar-java/src/org/keyczar/enums/CipherMode.java
/*
 * Copyright 2008 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.keyczar.enums;

import com.google.gson.annotations.Expose;
/**
 * Encodes different modes of operation:
 *   Cipher Block Chaining (CBC) with initial value (IV)
 *   Counter (CTR) with IV
 *   Electronic Code Book (ECB), no IV
 *   DET-CBC, CBC without IV
 *
 * @author steveweis@gmail.com (Steve Weis)
 *
 */
public enum CipherMode {
  CBC(0, "AES/CBC/PKCS5Padding", true),
  CTR(1, "AES/CTR/NoPadding", true),
  ECB(2, "AES/ECB/NoPadding", false),
  DET_CBC(3, "AES/CBC/PKCS5Padding", false);

  private String jceMode;
  @Expose
  private int value;

  private CipherMode(int v, String s, boolean useIv) {
    value = v;
    jceMode = s;
  }

  public String getMode() {
    return jceMode;
  }

  int getValue() {
    return value;
  }

  static CipherMode getMode(int value) {
    switch (value) {
      case 0:
        return CBC;
      case 1:
        return CTR;
      case 2:
        return ECB;
      case 3:
        return DET_CBC;
    }
    return null;
  }

  public int getOutputSize(int blockSize, int inputLength) {
    if (this == CBC) {
      return (inputLength / blockSize + 2) * blockSize;
    } else if (this == ECB) {
      return blockSize;
    } else if (this == CTR) {
      return inputLength + blockSize / 2;
    } else if (this == DET_CBC) {
      return (inputLength / blockSize + 1) * blockSize;
    } else {
      return 0;
    }
  }
}

File: cpp/src/keyczar/cipher_mode.cc
// Copyright 2009 Sebastien Martini (seb@dbzteam.org)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
#include <keyczar/cipher_mode.h>

#include <keyczar/base/logging.h>

namespace keyczar {

// static
CipherMode::Type CipherMode::GetTypeFromName(const std::string& name) {
  if (name == "CBC")
    return CBC;
  if (name == "CTR")
    return CTR;
  if (name == "ECB")
    return ECB;
  if (name == "DET_CBC")
    return DET_CBC;

  NOTREACHED();
  return UNDEF;
}

// static
std::string CipherMode::GetNameFromType(Type type) {
  switch (type) {
    case CBC:
      return "CBC";
    case CTR:
      return "CTR";
    case ECB:
      return "ECB";
    case DET_CBC:
      return "DET_CBC";
    default:
      NOTREACHED();
  }
  return "";
}

// static
bool CipherMode::HasIV(Type type) {
  switch (type) {
    case CBC:
    case CTR:
      return true;
    case ECB:
    case DET_CBC:
      return false;
    default:
      NOTREACHED();
  }
  return false;
}

// static
int CipherMode::GetOutputSize(Type type, int block_size, int input_length) {
  switch (type) {
    case CBC:
      return (input_length / block_size + 2) * block_size;
    case ECB:
      return block_size;
    case CTR:
      return input_length + block_size / 2;
    case DET_CBC:
      return (input_length / block_size + 1) * block_size;
    default:
      NOTREACHED();
  }
  return 0;
}

}  // namespace keyczar
